File: java/pl/edu/pg/mygame/organisms/animals/Human.java
package pl.edu.pg.mygame.organisms.animals;

import pl.edu.pg.mygame.Grid;
import pl.edu.pg.mygame.World;
import pl.edu.pg.mygame.organisms.Organism;
import pl.edu.pg.mygame.organisms.Species;

import java.awt.*;
import java.io.Serializable;
import java.util.List;

import static pl.edu.pg.mygame.utils.Constants.HumanActions.NO_MOVEMENT;
import static pl.edu.pg.mygame.utils.HelperFunctions.getRandomPoint;

public class Human extends Animal implements Serializable {
    private int action;
    private int last_ability;
    private boolean ability_on;
    public Human(Point p) {
        super(p);
        this.species = Species.HUMAN;
        this.strength = 5;
        this.initiative = 4;
        this.last_ability = -10;
        action = NO_MOVEMENT;
    }

    @Override
    protected Point getNewPoint(Grid grid) {
        return grid.getHumanPoint(position, action);
    }


    // here is only Alzur's shield implemented
    @Override
    public boolean collision(Organism other, boolean attacking, Grid grid) {
        if (!attacking && ability_on) {
            List<Point> neighbours = grid.getFreeNeighbourCells(other.getPosition());
            if (neighbours.size() > 0) {
                Point new_point = getRandomPoint(neighbours);
                World.message_panel.showMessage("Human deterred " + other.getSpecies() + " and it moved to: " +
                        String.format("(%d, %d)", new_point.x, new_point.y));
                grid.setOrganism(other.getPosition(), null);
                other.setPosition(new_point);
                grid.setOrganism(new_point, other);
            }
            else
                World.message_panel.showMessage("Human deterred " + other.getSpecies() + " and it didn't move (no free space)\n");
            return true;
        }
        return false;
    }


    public void setAction(int action, int curr_turn) {
        this.action = action;
        if (ability_on && curr_turn > last_ability + 5)
            ability_on = false;
    }

    public void activateAbility(int curr_turn) {
        if (last_ability + 10 <= curr_turn) {
            ability_on = true;
            last_ability = curr_turn;
            World.message_panel.showMessage("Human used ability Alzur's shield\n");
        }
    }
}

File: python/organisms/animals/human.py
from gui.logger import Logger
from organisms.animals.animal import Animal
from organisms.species import Species
from utils.helpers import Helpers


class Human(Animal):
    def __init__(self, pos):
        super().__init__(pos)
        self.strength = 5
        self.initiative = 4
        self.species = Species.HUMAN
        self.last_ability = -10
        self.action = 0
        self.new_position = self.position
        self.ability_on = False

    def update(self, key, grid, turn):
        self.new_position = grid.get_new_human_position(self.position, key)
        if self.ability_on and turn > self.last_ability + 5:
            self.ability_on = False

    def activate_ability(self, turn):
        if self.last_ability + 10 <= turn:
            self.ability_on = True
            self.last_ability = turn
            Logger.log("Human used ability Alzur's shield")

    def get_new_point(self, grid):
        return self.new_position

    def collision(self, other, attacking, grid):
        if not attacking and self.ability_on:
            neighbors = grid.get_free_neighbor_cells(other.position)
            if len(neighbors) > 0:
                new_point = Helpers.get_random_point(neighbors)
                Logger.log('Human deterred ' + other.species.name + ' and it moved to: ' + str(new_point))
                grid.set_organism(other.position, None)
                other.position = new_point
                grid.set_organism(new_point, other)
            else:
                Logger.log('Human deterred ' + other.species.name + " and it didn't move (no free space)")
            return True
        return False
